pkg/reconciler/trigger/controller.go

```diff
@@ -0,0 +1,137 @@
+
+	"knative.dev/eventing/pkg/apis/eventing"
+	eventingv1beta1 "knative.dev/eventing/pkg/apis/eventing/v1beta1"
+	"knative.dev/eventing/pkg/duck"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/pkg/client/injection/ducks/duck/v1/addressable"
+	"knative.dev/pkg/client/injection/ducks/duck/v1/conditions"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	pkgcontroller "knative.dev/pkg/controller"
+	pkgreconciler "knative.dev/pkg/reconciler"
+	"knative.dev/pkg/resolver"
+
+	brokerv1beta1 "github.com/google/knative-gcp/pkg/apis/broker/v1beta1"
+	brokerinformer "github.com/google/knative-gcp/pkg/client/injection/informers/broker/v1beta1/broker"
+	triggerinformer "github.com/google/knative-gcp/pkg/client/injection/informers/broker/v1beta1/trigger"
+	triggerreconciler "github.com/google/knative-gcp/pkg/client/injection/reconciler/broker/v1beta1/trigger"
+	metadataClient "github.com/google/knative-gcp/pkg/gclient/metadata"
+	"github.com/google/knative-gcp/pkg/reconciler"
+	"github.com/google/knative-gcp/pkg/utils"
+)
+
+const (
+	// controllerAgentName is the string used by this controller to identify
+	// itself when creating events.
+	controllerAgentName = "trigger-controller"
+	// finalizerName is the name of the finalizer that this controller adds to the Triggers that it reconciles.
+	finalizerName = "googlecloud"
+)
+
+// filterBroker is the function to filter brokers with proper brokerclass.
+var filterBroker = pkgreconciler.AnnotationFilterFunc(eventingv1beta1.BrokerClassAnnotationKey, brokerv1beta1.BrokerClass, false /*allowUnset*/)
+
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	// TODO initialize project ID here via env var or metadata.
+	projectID := ""
+
+	triggerInformer := triggerinformer.Get(ctx)
+
+	// Attempt to create a pubsub client for all worker threads to use. If this
+	// fails, pass a nil value to the Reconciler. They will attempt to
+	// create a client on reconcile.
+	client, err := newPubsubClient(ctx, projectID)
+	if err != nil {
+		logging.FromContext(ctx).Error("Failed to create controller-wide Pub/Sub client", zap.Error(err))
+	}
+
+	if client != nil {
+		go func() {
+			<-ctx.Done()
+			client.Close()
+		}()
+	}
+
+	r := &Reconciler{
+		Base:         reconciler.NewBase(ctx, controllerAgentName, cmw),
+		brokerLister: brokerinformer.Get(ctx).Lister(),
+		pubsubClient: client,
+		projectID:    projectID,
+	}
+
+	impl := triggerreconciler.NewImpl(ctx, r, withAgentAndFinalizer)
+	r.kresourceTracker = duck.NewListableTracker(ctx, conditions.Get, impl.EnqueueKey, controller.GetTrackerLease(ctx))
+	r.addressableTracker = duck.NewListableTracker(ctx, addressable.Get, impl.EnqueueKey, controller.GetTrackerLease(ctx))
+	r.uriResolver = resolver.NewURIResolver(ctx, impl.EnqueueKey)
+
+	r.Logger.Info("Setting up event handlers")
+
+	triggerInformer.Informer().AddEventHandlerWithResyncPeriod(controller.HandleAll(impl.Enqueue), reconciler.DefaultResyncPeriod)
```

Comment:
Do we need to filter on trigger label `eventing.BrokerLabelKey`?

Reply:
No because we need to be able to reconcile orphaned triggers without a broker.

Reply:
When would that happen? It seems trigger spec broker and broker class are both immutable.

Reply:
Sorry let me clarify. Do mean filter on the **existence** of the label or actually check the broker is GCP broker?

Let me assume you meant the latter . If you delete a broker, then its triggers become orphaned. We'd like to be able to still delete the triggers (in this case reconcile is a noop).